databuilder/query_model_transforms.py

```diff
@@ -0,0 +1,79 @@
+"""
+Apply modifications to the query graph which make it easier to work with or allow us to
+generate more efficient SQL
+"""
+import copy
+import dataclasses
+from typing import Any
+
+from databuilder.query_model import PickOneRowPerPatient, SelectColumn, get_input_nodes
+
+
+class PickOneRowPerPatientWithColumns(PickOneRowPerPatient):
+    selected_columns: frozenset[Any]
```

Comment:
What's the rationale for this class living here and not in `query_model.py`?

Reply:
Ah good question, and something I should have made more effort to argue for. As I see it, the only job of the query model itself is to capture the semantics of the query. It shouldn't be concerned with ease of expression or readability (which is ehrQL's job) or ease and efficiency of execution (which is the query engine's job). So these kinds of transformation to the query graph are a query engine implementation detail and shouldn't be mixed up in the core query model.

I anticipate us adding quite a few more of these kinds of query-engine-only node types in future to handle things like codelist tables.

Does that help? And do you agree? I'll expand the comments in the code if so.

Reply:
> Does that help? And do you agree?

yes * 2